### PR TITLE
feat: support the func concast_ws

### LINKED ISSUE #505: [SQL] Support MySQL `CONCAT_WS` function.
**What would you like to be added:**

**Why is this needed:**

ref #454 

official document:
https://dev.mysql.com/doc/refman/5.6/en/string-functions.html#function_concat-ws

code:
https://github.com/arana-db/arana/tree/master/pkg/runtime/function/scripts/string.js

reference code:
https://github.com/arana-db/arana/tree/master/pkg/runtime/function2/abs.go

### PR DESCRIPTION
**Pull request by Charlie17Li**
Signed-off-by: charlie <charlie1213800@gmail.com>

**What this PR does**:

support the func [concast_ws](https://dev.mysql.com/doc/refman/5.6/en/string-functions.html#function_concat-ws)

**Which issue(s) this PR fixes**:

Fixes #505 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/runtime/function2/concat_ws.go [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+	"strings"
+)
+
+import (
+	"github.com/arana-db/arana/pkg/proto"
+	"github.com/arana-db/arana/pkg/runtime/ast"
+)
+
+// FuncConcat_WS https://dev.mysql.com/doc/refman/5.6/en/string-functions.html#function_concat-ws
+const FuncConcat_WS = "CONCAT_WS"
```
Comment (jjeffcaii) [inline pkg/runtime/function2/concat_ws.go:33]: Rename to `FuncConcatWS`.

File: pkg/runtime/function2/concat_ws.go [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+		NullType = reflect.TypeOf(ast.Null{}).String()
+		sep      proto.Value
+		err      error
+	)
+
+	if sep, err = inputs[0].Value(ctx); err != nil {
+		return nil, err
+	}
+	// check separator is NULL
+	if strings.Compare(reflect.TypeOf(sep).String(), NullType) == 0 {
```
Comment (jjeffcaii) [inline pkg/runtime/function2/concat_ws.go:55]: I think `_,isNull := sep.(ast.Null)` is enough.

**Review by jjeffcaii — COMMENTED** [review]
pls fix

**Review by dongzl — APPROVED** [review]

File: pkg/runtime/function2/concat_ws.go [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+type concatWSFunc struct{}
+
+func (c concatWSFunc) Apply(ctx context.Context, inputs ...proto.Valuer) (proto.Value, error) {
+	var (
+		sb  strings.Builder
+		sep proto.Value
+		err error
+	)
+
+	if sep, err = inputs[0].Value(ctx); err != nil {
```
[the remark this answers is not in the document]
Comment (Charlie17Li) [reply pkg/runtime/function2/concat_ws.go:49]: function 有一个`NumInput`方法 申明了需要的参数数量，上游貌似应该是有做判断，我去确认下
https://github.com/arana-db/arana/pull/506/files#diff-8e8665f5fdb836007525ba20aa30dba6d2dad351c67eb11bcc4ad308efe92e86R76-R78

[the remark this answers is not in the document]
Comment (jjeffcaii) [reply pkg/runtime/function2/concat_ws.go:49]: NumInput会自动判断最小需要的参数数量，所以这里不需要再次判断了。

**Review by jjeffcaii — APPROVED** [review]

**Pull request merged**